// @sentry/electron/main/integrations/child-process.js
const core = require('@sentry/core');
const electron = require('electron');
const electronNormalize = require('../electron-normalize.js');

const DEFAULT_OPTIONS = {
    breadcrumbs: electronNormalize.EXIT_REASONS,
    events: ['abnormal-exit', 'launch-failed', 'integrity-failure'],
};
/** Gets message and severity */
function getMessageAndSeverity(reason, proc) {
    const message = `'${proc}' process exited with '${reason}'`;
    switch (reason) {
        case 'abnormal-exit':
        case 'killed':
            return { message, level: 'warning' };
        case 'crashed':
        case 'oom':
        case 'launch-failed':
        case 'integrity-failure':
            return { message, level: 'fatal' };
        default:
            return { message, level: 'debug' };
    }
}
/**
 * Adds breadcrumbs for Electron child process events.
 */
const childProcessIntegration = core.defineIntegration((userOptions = {}) => {
    const { breadcrumbs, events } = userOptions;
    const options = {
        breadcrumbs: Array.isArray(breadcrumbs) ? breadcrumbs : breadcrumbs === false ? [] : DEFAULT_OPTIONS.breadcrumbs,
        events: Array.isArray(events) ? events : events === false ? [] : DEFAULT_OPTIONS.events,
    };
    return {
        name: 'ChildProcess',
        setup(client) {
            const { breadcrumbs, events } = options;
            const allReasons = Array.from(new Set([...breadcrumbs, ...events]));
            // only hook these events if we're after more than just the unresponsive event
            if (allReasons.length > 0) {
                const clientOptions = client.getOptions();
                electron.app.on('child-process-gone', (_, details) => {
                    const { reason } = details;
                    // Capture message first
                    if (events.includes(reason)) {
                        const { message, level } = getMessageAndSeverity(details.reason, details.type);
                        core.captureMessage(message, { level, tags: { 'event.process': details.type } });
                    }
                    // And then add breadcrumbs for subsequent events
                    if (breadcrumbs.includes(reason)) {
                        core.addBreadcrumb({
                            type: 'process',
                            category: 'child-process',
                            ...getMessageAndSeverity(details.reason, details.type),
                            data: details,
                        });
                    }
                });
                electron.app.on('render-process-gone', (_, contents, details) => {
                    const { reason } = details;
                    const name = clientOptions?.getRendererName?.(contents) || 'renderer';
                    // Capture message first
                    if (events.includes(reason)) {
                        const { message, level } = getMessageAndSeverity(details.reason, name);
                        core.captureMessage(message, level);
                    }
                    // And then add breadcrumbs for subsequent events
                    if (breadcrumbs.includes(reason)) {
                        core.addBreadcrumb({
                            type: 'process',
                            category: 'child-process',
                            ...getMessageAndSeverity(details.reason, name),
                            data: details,
                        });
                    }
                });
            }
        },
    };
});

exports.childProcessIntegration = childProcessIntegration;//# sourceMappingURL=https://cursor-sourcemaps.s3.amazonaws.com/sourcemaps/7801a556824585b7f2721900066bc87c4a09b740/node_modules/@sentry/electron/main/integrations/child-process.js.map
