// @sentry/electron/renderer/metrics.js
const core = require('@sentry/core');
const ipc = require('./ipc.js');

/**
 * Sends metrics to the Electron main process where they can be aggregated in a single process
 */
class ElectronRendererMetricsAggregator {
    constructor() {
        this._ipc = ipc.getIPC();
    }
    /** @inheritdoc */
    add(metricType, name, value, unit, tags, timestamp) {
        this._ipc.sendAddMetric({ metricType, name, value, unit, tags, timestamp });
    }
    /** @inheritdoc */
    flush() {
        // do nothing
    }
    /** @inheritdoc */
    close() {
        // do nothing
    }
    /** @inheritdoc */
    toString() {
        return '';
    }
}
/**
 * Adds a value to a counter metric
 *
 * @experimental This API is experimental and might have breaking changes in the future.
 */
function increment(name, value = 1, data) {
    core.metrics.increment(ElectronRendererMetricsAggregator, name, value, data);
}
/**
 * Adds a value to a distribution metric
 *
 * @experimental This API is experimental and might have breaking changes in the future.
 */
function distribution(name, value, data) {
    core.metrics.distribution(ElectronRendererMetricsAggregator, name, value, data);
}
/**
 * Adds a value to a set metric. Value must be a string or integer.
 *
 * @experimental This API is experimental and might have breaking changes in the future.
 */
function set(name, value, data) {
    core.metrics.set(ElectronRendererMetricsAggregator, name, value, data);
}
/**
 * Adds a value to a gauge metric
 *
 * @experimental This API is experimental and might have breaking changes in the future.
 */
function gauge(name, value, data) {
    core.metrics.gauge(ElectronRendererMetricsAggregator, name, value, data);
}
function timing(name, value, unit = 'second', data) {
    core.metrics.timing(ElectronRendererMetricsAggregator, name, value, unit, data);
}
const metrics = {
    increment,
    distribution,
    set,
    gauge,
    timing,
};

exports.metrics = metrics;//# sourceMappingURL=https://cursor-sourcemaps.s3.amazonaws.com/sourcemaps/7801a556824585b7f2721900066bc87c4a09b740/node_modules/@sentry/electron/renderer/metrics.js.map
