// @sentry/electron/main/index.js
const node = require('@sentry/node');
const electronBreadcrumbs = require('./integrations/electron-breadcrumbs.js');
const onuncaughtexception = require('./integrations/onuncaughtexception.js');
const index = require('./integrations/sentry-minidump/index.js');
const electronMinidump = require('./integrations/electron-minidump.js');
const preloadInjection = require('./integrations/preload-injection.js');
const mainProcessSession = require('./integrations/main-process-session.js');
const browserWindowSession = require('./integrations/browser-window-session.js');
const additionalContext = require('./integrations/additional-context.js');
const netBreadcrumbs = require('./integrations/net-breadcrumbs.js');
const childProcess = require('./integrations/child-process.js');
const screenshots = require('./integrations/screenshots.js');
const rendererProfiling = require('./integrations/renderer-profiling.js');
const normalizePaths = require('./integrations/normalize-paths.js');
const electronContext = require('./integrations/electron-context.js');
const anr = require('./integrations/anr.js');
const electronNet = require('./transports/electron-net.js');
const electronOfflineNet = require('./transports/electron-offline-net.js');
const sdk = require('./sdk.js');
const stackParse = require('./stack-parse.js');
const ipc = require('../common/ipc.js');



exports.DEFAULT_USER_INCLUDES = node.DEFAULT_USER_INCLUDES;
exports.NodeClient = node.NodeClient;
exports.SEMANTIC_ATTRIBUTE_SENTRY_OP = node.SEMANTIC_ATTRIBUTE_SENTRY_OP;
exports.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN = node.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN;
exports.SEMANTIC_ATTRIBUTE_SENTRY_SAMPLE_RATE = node.SEMANTIC_ATTRIBUTE_SENTRY_SAMPLE_RATE;
exports.SEMANTIC_ATTRIBUTE_SENTRY_SOURCE = node.SEMANTIC_ATTRIBUTE_SENTRY_SOURCE;
exports.Scope = node.Scope;
exports.SentryContextManager = node.SentryContextManager;
exports.addBreadcrumb = node.addBreadcrumb;
exports.addEventProcessor = node.addEventProcessor;
exports.addIntegration = node.addIntegration;
exports.addOpenTelemetryInstrumentation = node.addOpenTelemetryInstrumentation;
exports.addRequestDataToEvent = node.addRequestDataToEvent;
exports.amqplibIntegration = node.amqplibIntegration;
exports.captureCheckIn = node.captureCheckIn;
exports.captureConsoleIntegration = node.captureConsoleIntegration;
exports.captureEvent = node.captureEvent;
exports.captureException = node.captureException;
exports.captureFeedback = node.captureFeedback;
exports.captureMessage = node.captureMessage;
exports.captureSession = node.captureSession;
exports.close = node.close;
exports.connectIntegration = node.connectIntegration;
exports.consoleIntegration = node.consoleIntegration;
exports.contextLinesIntegration = node.contextLinesIntegration;
exports.continueTrace = node.continueTrace;
exports.createGetModuleFromFilename = node.createGetModuleFromFilename;
exports.createTransport = node.createTransport;
exports.cron = node.cron;
exports.dataloaderIntegration = node.dataloaderIntegration;
exports.debugIntegration = node.debugIntegration;
exports.dedupeIntegration = node.dedupeIntegration;
exports.endSession = node.endSession;
exports.expressErrorHandler = node.expressErrorHandler;
exports.expressIntegration = node.expressIntegration;
exports.extraErrorDataIntegration = node.extraErrorDataIntegration;
exports.extractRequestData = node.extractRequestData;
exports.fastifyIntegration = node.fastifyIntegration;
exports.flush = node.flush;
exports.fsIntegration = node.fsIntegration;
exports.functionToStringIntegration = node.functionToStringIntegration;
exports.generateInstrumentOnce = node.generateInstrumentOnce;
exports.genericPoolIntegration = node.genericPoolIntegration;
exports.getActiveSpan = node.getActiveSpan;
exports.getAutoPerformanceIntegrations = node.getAutoPerformanceIntegrations;
exports.getClient = node.getClient;
exports.getCurrentHub = node.getCurrentHub;
exports.getCurrentScope = node.getCurrentScope;
exports.getGlobalScope = node.getGlobalScope;
exports.getIsolationScope = node.getIsolationScope;
exports.getRootSpan = node.getRootSpan;
exports.getSpanDescendants = node.getSpanDescendants;
exports.getSpanStatusFromHttpCode = node.getSpanStatusFromHttpCode;
exports.getTraceData = node.getTraceData;
exports.getTraceMetaTags = node.getTraceMetaTags;
exports.graphqlIntegration = node.graphqlIntegration;
exports.hapiIntegration = node.hapiIntegration;
exports.httpIntegration = node.httpIntegration;
exports.inboundFiltersIntegration = node.inboundFiltersIntegration;
exports.initOpenTelemetry = node.initOpenTelemetry;
exports.isInitialized = node.isInitialized;
exports.kafkaIntegration = node.kafkaIntegration;
exports.koaIntegration = node.koaIntegration;
exports.lastEventId = node.lastEventId;
exports.linkedErrorsIntegration = node.linkedErrorsIntegration;
exports.localVariablesIntegration = node.localVariablesIntegration;
exports.lruMemoizerIntegration = node.lruMemoizerIntegration;
exports.metrics = node.metrics;
exports.modulesIntegration = node.modulesIntegration;
exports.mongoIntegration = node.mongoIntegration;
exports.mongooseIntegration = node.mongooseIntegration;
exports.mysql2Integration = node.mysql2Integration;
exports.mysqlIntegration = node.mysqlIntegration;
exports.nativeNodeFetchIntegration = node.nativeNodeFetchIntegration;
exports.nestIntegration = node.nestIntegration;
exports.nodeContextIntegration = node.nodeContextIntegration;
exports.onUnhandledRejectionIntegration = node.onUnhandledRejectionIntegration;
exports.parameterize = node.parameterize;
exports.postgresIntegration = node.postgresIntegration;
exports.prismaIntegration = node.prismaIntegration;
exports.profiler = node.profiler;
exports.redisIntegration = node.redisIntegration;
exports.requestDataIntegration = node.requestDataIntegration;
exports.rewriteFramesIntegration = node.rewriteFramesIntegration;
exports.sessionTimingIntegration = node.sessionTimingIntegration;
exports.setContext = node.setContext;
exports.setCurrentClient = node.setCurrentClient;
exports.setExtra = node.setExtra;
exports.setExtras = node.setExtras;
exports.setHttpStatus = node.setHttpStatus;
exports.setMeasurement = node.setMeasurement;
exports.setNodeAsyncContextStrategy = node.setNodeAsyncContextStrategy;
exports.setTag = node.setTag;
exports.setTags = node.setTags;
exports.setUser = node.setUser;
exports.setupConnectErrorHandler = node.setupConnectErrorHandler;
exports.setupExpressErrorHandler = node.setupExpressErrorHandler;
exports.setupFastifyErrorHandler = node.setupFastifyErrorHandler;
exports.setupHapiErrorHandler = node.setupHapiErrorHandler;
exports.setupKoaErrorHandler = node.setupKoaErrorHandler;
exports.setupNestErrorHandler = node.setupNestErrorHandler;
exports.spanToBaggageHeader = node.spanToBaggageHeader;
exports.spanToJSON = node.spanToJSON;
exports.spanToTraceHeader = node.spanToTraceHeader;
exports.spotlightIntegration = node.spotlightIntegration;
exports.startInactiveSpan = node.startInactiveSpan;
exports.startNewTrace = node.startNewTrace;
exports.startSession = node.startSession;
exports.startSpan = node.startSpan;
exports.startSpanManual = node.startSpanManual;
exports.suppressTracing = node.suppressTracing;
exports.trpcMiddleware = node.trpcMiddleware;
exports.validateOpenTelemetrySetup = node.validateOpenTelemetrySetup;
exports.withActiveSpan = node.withActiveSpan;
exports.withIsolationScope = node.withIsolationScope;
exports.withMonitor = node.withMonitor;
exports.withScope = node.withScope;
exports.zodErrorsIntegration = node.zodErrorsIntegration;
exports.electronBreadcrumbsIntegration = electronBreadcrumbs.electronBreadcrumbsIntegration;
exports.onUncaughtExceptionIntegration = onuncaughtexception.onUncaughtExceptionIntegration;
exports.sentryMinidumpIntegration = index.sentryMinidumpIntegration;
exports.electronMinidumpIntegration = electronMinidump.electronMinidumpIntegration;
exports.preloadInjectionIntegration = preloadInjection.preloadInjectionIntegration;
exports.mainProcessSessionIntegration = mainProcessSession.mainProcessSessionIntegration;
exports.browserWindowSessionIntegration = browserWindowSession.browserWindowSessionIntegration;
exports.additionalContextIntegration = additionalContext.additionalContextIntegration;
exports.electronNetIntegration = netBreadcrumbs.electronNetIntegration;
exports.childProcessIntegration = childProcess.childProcessIntegration;
exports.screenshotsIntegration = screenshots.screenshotsIntegration;
exports.rendererProfileFromIpc = rendererProfiling.rendererProfileFromIpc;
exports.normalizePathsIntegration = normalizePaths.normalizePathsIntegration;
exports.electronContextIntegration = electronContext.electronContextIntegration;
exports.anrIntegration = anr.anrIntegration;
exports.makeElectronTransport = electronNet.makeElectronTransport;
exports.makeElectronOfflineTransport = electronOfflineNet.makeElectronOfflineTransport;
exports.getDefaultIntegrations = sdk.getDefaultIntegrations;
exports.init = sdk.init;
exports.defaultStackParser = stackParse.defaultStackParser;
Object.defineProperty(exports, "IPCMode", {
	enumerable: true,
	get: () => ipc.IPCMode
});//# sourceMappingURL=https://cursor-sourcemaps.s3.amazonaws.com/sourcemaps/7801a556824585b7f2721900066bc87c4a09b740/node_modules/@sentry/electron/main/index.js.map
